src/main/java/org/radarcns/biovotion/BiovotionDeviceManager.java

```diff
@@ -317,6 +408,206 @@ public void onScanError(@NonNull Scanner scanner, @NonNull BleScanException thro
+         * - the current GAP status is queried, if a GAP request is running (=0) the new request is aborted
+         * - the number of records and latest counter value are queried
+         * - the new starting index is the latest counter value
+         * - the range of values to request is calculated via the difference of the current latest counter and the latest counter at time of the last GAP request
+         * - a new GAP request is issued with the calculated values, and the current latest counter value is saved
+         *
+         * Pages:
+         * The data on the device is stored in pages of a specific number of samples (e.g. 17 in case of raw data). When getting data via GAP request,
+         * data will always be streamed in whole pages, i.e. if the start or end point of a GAP request is in the middle of a page, that whole page will
+         * still be streamed. This may result in more data being streamed than was requested.
+         *
+         */
+
+        // read GAP request status
+        else if (p.id() == VsmConstants.PID_GAP_REQUEST_STATUS) {
+            gap_stat = p.value()[0];
+            if (gap_stat == 0) {
+                logger.info("Biovotion VSM GAP request running");
+                return;
+            }
+            // TODO: handle GAP error statuses (gap_stat > 1)
+
+            // will always stream whole pages, don't even bother if stream is mid page
+            //if (gap_raw_since_last % VsmConstants.GAP_MAX_PER_PAGE_VITAL_RAW != 0)
+            //    return;
+
+            logger.info("Biovotion VSM GAP status: raw_cnt:{} | raw_num:{} | gap_stat:{}", gap_raw_cnt, gap_raw_num, gap_stat);
+
+            paramReadRequest(VsmConstants.PID_NUMBER_OF_RAW_DATA_SETS_IN_STORAGE);
+        }
+
+        // read number of raw data sets currently in device storage
+        else if (p.id() == VsmConstants.PID_NUMBER_OF_RAW_DATA_SETS_IN_STORAGE) {
+            gap_raw_num = p.valueAsInteger();
+            logger.info("Biovotion VSM GAP status: raw_cnt:{} | raw_num:{} | gap_stat:{}", gap_raw_cnt, gap_raw_num, gap_stat);
+
+            paramReadRequest(VsmConstants.PID_LAST_RAW_COUNTER_VALUE);
+        }
+
+        // read current latest record counter id. may initiate GAP request here
+        else if (p.id() == VsmConstants.PID_LAST_RAW_COUNTER_VALUE) {
+            gap_raw_cnt = p.valueAsInteger();
+            if (gap_raw_last_cnt < 0) gap_raw_last_cnt = gap_raw_cnt; // init
+
+            int new_idx = gap_raw_cnt;
+
+            // set the number of records to request at this point
+            //int records_to_get = VsmConstants.GAP_NUM_PAGES * VsmConstants.GAP_MAX_PER_PAGE_VITAL_RAW;
+            int records_to_get = gap_raw_cnt - gap_raw_last_cnt;
+
+            logger.info("Biovotion VSM GAP status: raw_cnt:{}:{} | raw_toget:{}:{} | gap_stat:{}", gap_raw_cnt, gap_raw_last_cnt, records_to_get, gap_raw_to_get, gap_stat);
```

Comment:
Yeah since I use logcat (android studio) slf4j debug doesn't show up so I just use info everywhere, but I'll see that I set them to debug where applicable. Any tips on getting logcat to show logger.debug?